server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java

```diff
@@ -30,90 +27,88 @@ public class UpstreamRouteProvider {
-
     /**
-     * Cached load balancers for config deployments
+     * Maximum idle period while balancers will stay in the local cache.
      */
-    private volatile Map<String, TieredBalancer> balancers = new HashMap<>();
+    private static final long IDLE_PERIOD_IN_MS = TimeUnit.HOURS.toMillis(1);
+
 
     /**
-     * Returns UpstreamRoute for the given provider
-     *
-     * @param provider upstream provider for any deployment with actual upstreams
-     * @return upstream route
+     * Cached load balancers
      */
-    public UpstreamRoute get(UpstreamProvider provider) {
-        String deploymentName = provider.getName();
-        List<Upstream> upstreams = provider.getUpstreams();
-
-        TieredBalancer balancer = balancers.get(deploymentName);
-        if (balancer == null) {
-            // if no state found for upstream, it's probably custom application
-            balancer = new TieredBalancer(deploymentName, upstreams);
-        }
+    private final ConcurrentHashMap<String, TieredBalancer> balancers = new ConcurrentHashMap<>();
 
-        return new UpstreamRoute(balancer, MAX_RETRY_COUNT);
+    public UpstreamRouteProvider(Vertx vertx) {
+        vertx.setPeriodic(0, TimeUnit.MINUTES.toMillis(1), event -> evictExpiredBalancers());
     }
 
-    public synchronized void onUpdate(Config config) {
-        log.debug("Updating load balancers state");
-        Map<String, TieredBalancer> oldState = balancers;
-        Map<String, TieredBalancer> newState = new HashMap<>();
-
-        Map<String, Model> models = config.getModels();
-        updateDeployments(newState, oldState, models.values());
-
-        Map<String, Application> applications = config.getApplications();
-        updateDeployments(newState, oldState, applications.values());
-
-        Map<String, Addon> addons = config.getAddons();
-        updateDeployments(newState, oldState, addons.values());
-
-        Map<String, Assistant> assistants = config.getAssistant().getAssistants();
-        updateDeployments(newState, oldState, assistants.values());
-
-        LinkedHashMap<String, Route> routes = config.getRoutes();
-        updateRoutes(newState, oldState, routes.values());
+    public UpstreamRoute get(Deployment deployment) {
+        String key = getKey(deployment);
+        List<Upstream> upstreams = getUpstreams(deployment);
+        return get(key, upstreams);
+    }
 
-        balancers = newState;
+    public UpstreamRoute get(Route route) {
+        String key = getKey(route);
+        return get(key, route.getUpstreams());
     }
 
-    private static void updateRoutes(Map<String, TieredBalancer> newState, Map<String, TieredBalancer> oldState, Collection<Route> routes) {
-        for (Route route : routes) {
-            String name = route.getName();
+    private UpstreamRoute get(String key, List<Upstream> upstreams) {
+        TieredBalancer balancer = balancers.merge(key, new TieredBalancer(key, upstreams), (prev, next) -> {
```

Comment:
use compute to not create TieredBalancer all the time, it is quite expensive

Reply:
fixed